### PR TITLE
Share project-aware call resolution

### LINKED ISSUE #92: [Feature]: Reuse project-aware type and call resolution across rules
## Status Quo

PR #91 added project-wide `unused-decl` precision for build root discovery, import-aware public API surfaces, typed receiver calls, and result-location method calls. That machinery currently lives inside `src/project_unused_decl.zig`, so other rules cannot reuse it directly.

Other rules and checkers already have overlapping local logic:

- `store-violations-engine` and `stack-escape-engine` do syntactic FQN/resource call matching.
- `sentinel-alloc` has separate result-context type lookup.
- `deinit-lifecycle` only handles simple identifier receivers.
- `identifier-style`, per-file `unused-decl`, and `return-local-ptr` rely on local heuristics for aliases, type-like values, and pointer-like returns.

## Objectives

Make the resolution improvements from project-wide `unused-decl` available to other rules without duplicating AST decoding. Users should see fewer false positives and fewer missed diagnostics when code uses imported file structs, aliases, typed receivers, and result-location method calls. The feature should improve precision while keeping rule behavior conservative.

## User Flow

**Trigger:** The user runs zwanzig normally on a Zig project.

1. Zwanzig analyzes files and creates per-file type context as it does today.
2. Rules/checkers ask shared resolver helpers for import targets, receiver types, FQNs, expected result types, and method call identity.
3. Rules use the resolved information to avoid name-only assumptions and handle common Zig idioms.
4. Diagnostics remain stable, deterministic, and suppressible.

**Result:** Existing rules produce more accurate findings on real Zig codebases without requiring users to enable a separate mode.

## Scope

**In scope:**

- Extract reusable import/path resolution from `project_unused_decl.zig`.
- Extract reusable call resolution for identifier calls, field calls, FQNs, receiver types, and result-location method calls.
- Extend `TypeContext` or a related analysis helper so rules do not maintain separate local type resolvers.
- Apply the shared machinery to the highest-value consumers:
  - `store-violations-engine`
  - `stack-escape-engine`
  - `sentinel-alloc`
  - `deinit-lifecycle`
- Add targeted fixtures for imported file structs, typed receivers, aliases, and result-location calls.

**Out of scope:**

- Full Zig semantic analysis or cross-package compilation.
- Replacing ZIR/type context wholesale.
- Broad behavior changes to unrelated local syntax rules like `todo`, `empty-defer`, or `shadowed-variable`.
- Making every suggested consumer perfect in one PR if the shared extraction reveals too much coupling.

### Affected Modules

- `src/analysis/import_resolver.zig`: New shared import/path/package-root resolution helpers.
- `src/analysis/call_resolver.zig`: New shared call identity, receiver, FQN, and result-location method helpers.
- `src/type_context.zig`: Add or expose scoped type queries needed by call/result-location resolution.
- `src/project_unused_decl.zig`: Replace private duplicated helpers with shared analysis helpers.
- `src/checkers/store_violations_engine.zig` and `src/engine/analysis/resource_calls.zig`: Use shared call resolution for resource model matching.
- `src/checkers/stack_escape_engine.zig`: Use shared call resolution for escape model and thread spawn matching.
- `src/rules/sentinel_alloc.zig`: Use shared result-context/type resolution.
- `src/rules/deinit_lifecycle.zig`: Use shared receiver/call resolution for non-trivial receivers.
- `test/fixtures/*`: Add rule-specific fixtures.
- `docs/RULES.md` and `CHANGELOG.md`: Document user-visible precision improvements.

### Tasks

Each task is independently testable. Ordered by implementation sequence.

1. [ ] Extract import/path helpers from `project_unused_decl.zig` — `src/analysis/import_resolver.zig`
2. [ ] Extract call receiver/FQN/result-location helpers — `src/analysis/call_resolver.zig`
3. [ ] Refactor project-wide `unused-decl` to use the shared helpers without changing diagnostics — `src/project_unused_decl.zig`
4. [ ] Improve resource and escape model matching with shared call resolution — `src/engine/analysis/resource_calls.zig`, `src/checkers/stack_escape_engine.zig`
5. [ ] Improve result-context detection for sentinel allocations — `src/rules/sentinel_alloc.zig`
6. [ ] Improve cleanup receiver detection for lifecycle checks — `src/rules/deinit_lifecycle.zig`
7. [ ] Write tests: imported file structs, aliases, typed receivers, chained receivers, result-location factories, and no-regression fixtures for false positives
8. [ ] Update docs/PRD.md (if present)
9. [ ] Update docs/TRACEABILITY.md (if present)
10. [ ] Update docs/ARCHITECTURE.md (if architecture changed and file present)
11. [ ] Update `docs/RULES.md` and `CHANGELOG.md`

### New Dependencies

None.

## Acceptance Criteria

- [ ] All tasks completed
- [ ] Tests cover happy path and documented edge cases
- [ ] docs/PRD.md updated (if present)
- [ ] docs/TRACEABILITY.md updated (if present)
- [ ] Pre-commit hooks pass
- [ ] `just test` passes
- [ ] `just lint` passes
- [ ] Existing project-wide `unused-decl` fixture behavior remains unchanged except for intentional precision improvements
- [ ] At least one real Zig codebase spot check shows no obvious new false-positive class from the refactor

### PR DESCRIPTION
**Pull request by forketyfork**
## Summary

Extracts shared project-aware import and call resolution helpers and reuses them from project-wide unused declarations, resource calls, stack escape, sentinel allocation, and deinit lifecycle checks.

Adds fixtures for identifier resource models, typed escape models, result-location sentinel preservation, and chained cleanup receiver detection.

Fixes #92

### CODE REVIEW

File: src/rules/deinit_lifecycle.zig [hunk 1]
```diff
@@ -237,13 +237,11 @@ pub const DeinitLifecycleRule = struct {
...
-            if (receiver_node == 0 or receiver_node >= self.tags.len or self.tags[receiver_node] != .identifier) return null;
-
-            const receiver_token = self.tree.nodes.items(.main_token)[receiver_node];
-            if (receiver_token >= token_tags.len or token_tags[receiver_token] != .identifier) return null;
+            if (receiver_node == 0 or receiver_node >= self.tags.len) return null;
+            const receiver = call_utils.receiverSourceSlice(self.src.getContent(), self.tree, receiver_node) orelse return null;
 
             return .{
-                .receiver = self.tree.tokenSlice(receiver_token),
+                .receiver = receiver,
```
[the remark this answers is not in the document]
Comment (forketyfork) [reply src/rules/deinit_lifecycle.zig:249]: Fixed in `5a6425a`. `deinit-lifecycle` now canonicalizes identifier and field-access receiver chains before storing or comparing them, so formatting differences like `holder.value` versus `holder . value` no longer split the same cleanup receiver.

File: src/engine/analysis/resource_calls.zig [hunk 2]
```diff
@@ -148,19 +126,21 @@ pub fn mixin(comptime _Engine: type) type {
...
+                        return .{ .kind = .open, .target_expr = null, .call_node = call_ast_node };
+                    }
                 }
             }
 
-            if (std.mem.eql(u8, field_name, "close")) {
-                const target_expr = first_arg orelse base_node;
+            if (std.mem.eql(u8, call_info.method_name, "close")) {
+                const target_expr: ?u32 = if (first_arg) |arg| arg else call_info.base_node;
                 return .{ .kind = .close, .target_expr = target_expr, .call_node = call_ast_node };
```
[the remark this answers is not in the document]
Comment (forketyfork) [reply src/engine/analysis/resource_calls.zig]: Fixed in `5a6425a`. The built-in `close` fallback now only treats receiver calls like `file.close()` as closing that receiver, with a narrow exception for known `std.posix.close(fd)` calls. Bare identifier calls such as `close(file)` are no longer treated as a close unless a user config resource model explicitly says so.

File: src/rules/deinit_lifecycle.zig [hunk 3]
```diff
@@ -313,12 +311,10 @@ pub const DeinitLifecycleRule = struct {
...
             const rhs = @intFromEnum(pair[1]);
-            if (lhs == 0 or lhs >= self.tags.len or self.tags[lhs] != .identifier) return null;
+            if (lhs == 0 or lhs >= self.tags.len) return null;
 
-            const token_tags = self.tree.tokens.items(.tag);
-            const lhs_token = self.tree.nodes.items(.main_token)[lhs];
-            if (lhs_token >= token_tags.len or token_tags[lhs_token] != .identifier) return null;
-            if (!std.mem.eql(u8, self.tree.tokenSlice(lhs_token), receiver)) return null;
+            const lhs_receiver = call_utils.receiverSourceSlice(self.src.getContent(), self.tree, lhs) orelse return null;
+            if (!std.mem.eql(u8, lhs_receiver, receiver)) return null;
```
[the remark this answers is not in the document]
Comment (forketyfork) [reply src/rules/deinit_lifecycle.zig]: Fixed in `5a6425a`. Try-reinit matching now uses the same canonical receiver chain as cleanup tracking, so assignment LHS matching is no longer byte-for-byte source-slice sensitive.
